File: projects/urlshortener/bench/seed.ts
import 'dotenv/config';

const BASE_URL = process.env.BASE_URL || 'http://localhost:8080';
const COUNT = Number(process.env.COUNT) || 1000;
const OUTPUT = process.env.OUTPUT || 'codes.json';

interface CreateResponse {
  short_code: string;
  short_url: string;
  original_url: string;
}

async function createUrl(index: number): Promise<string | null> {
  const res = await fetch(`${BASE_URL}/api/v1/urls`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ url: `https://example.com/seed/${index}` }),
  });

  if (!res.ok) {
    console.error(`Failed to create URL ${index}: ${res.status}`);
    return null;
  }

  const data: CreateResponse = await res.json();
  return data.short_code;
}

async function main() {
  console.log(`Seeding ${COUNT} URLs to ${BASE_URL}...`);

  const codes: string[] = [];
  const batchSize = 50;

  for (let i = 0; i < COUNT; i += batchSize) {
    const batch = Array.from({ length: Math.min(batchSize, COUNT - i) }, (_, j) =>
      createUrl(i + j)
    );
    const results = await Promise.all(batch);
    codes.push(...results.filter((c): c is string => c !== null));
    console.log(`Progress: ${codes.length}/${COUNT}`);
  }

  await Bun.write(OUTPUT, JSON.stringify(codes, null, 2));
  console.log(`Saved ${codes.length} codes to ${OUTPUT}`);
}

main();

File: projects/urlshortener/bench/internal/seed/seed.go
package seed

import (
	"bytes"
	"context"
	"crypto/tls"
	"encoding/json"
	"fmt"
	"net/http"
	"runtime"
	"sync/atomic"
	"time"

	"golang.org/x/sync/errgroup"
)

type batchRequest struct {
	URLs []string `json:"urls"`
}

type createResponse struct {
	ShortCode string `json:"short_code"`
}

type batchResponse struct {
	URLs []createResponse `json:"urls"`
}

const bypassHeader = "X-Rate-Limit-Bypass"

func Run(baseURL string, count, batchSize int, bypassSecret string, insecureSkipVerify bool, timeout time.Duration) ([]string, error) {
	numWorkers := runtime.NumCPU() * 2
	fmt.Printf("Seeding %d URLs (batch size: %d, workers: %d)...\n", count, batchSize, numWorkers)

	client := &http.Client{
		Timeout: timeout,
		Transport: &http.Transport{
			TLSClientConfig:     &tls.Config{InsecureSkipVerify: insecureSkipVerify},
			MaxIdleConns:        numWorkers * 2,
			MaxIdleConnsPerHost: numWorkers * 2,
			IdleConnTimeout:     90 * time.Second,
			ForceAttemptHTTP2:   true,
		},
	}

	numBatches := (count + batchSize - 1) / batchSize
	results := make([][]string, numBatches)
	var progress atomic.Int64

	g, _ := errgroup.WithContext(context.Background())
	g.SetLimit(numWorkers)

	for i := range numBatches {
		batchIndex := i
		startIndex := batchIndex * batchSize
		currentBatch := min(batchSize, count-startIndex)

		g.Go(func() error {
			batchCodes, err := createBatch(client, baseURL, startIndex, currentBatch, bypassSecret)
			if err != nil {
				return fmt.Errorf("failed to create batch at %d: %w", startIndex, err)
			}
			results[batchIndex] = batchCodes
			done := progress.Add(int64(len(batchCodes)))
			fmt.Printf("\rProgress: %d/%d", done, count)
			return nil
		})
	}

	if err := g.Wait(); err != nil {
		return nil, err
	}

	codes := make([]string, 0, count)
	for _, batch := range results {
		codes = append(codes, batch...)
	}

	fmt.Printf("\nSeeding complete: %d codes\n", len(codes))
	return codes, nil
}

func createBatch(client *http.Client, baseURL string, startIndex, count int, bypassSecret string) ([]string, error) {
	urls := make([]string, count)
	for i := range count {
		urls[i] = fmt.Sprintf("https://example.com/seed/%d", startIndex+i)
	}

	body, err := json.Marshal(batchRequest{URLs: urls})
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequest(http.MethodPost, baseURL+"/api/v1/urls/batch", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	if bypassSecret != "" {
		req.Header.Set(bypassHeader, bypassSecret)
	}

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusCreated {
		return nil, fmt.Errorf("unexpected status: %d", resp.StatusCode)
	}

	var result batchResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	codes := make([]string, len(result.URLs))
	for i, u := range result.URLs {
		codes[i] = u.ShortCode
	}
	return codes, nil
}
